File: spark-doris-connector/spark-doris-connector-base/src/main/scala/org/apache/doris/spark/client/read/DorisThriftReader.scala
package org.apache.doris.spark.client.read

import org.apache.doris.spark.client.{DorisReaderPartition, DorisSchema}
import org.apache.doris.spark.config.{DorisConfig, DorisOptions}
import org.apache.doris.spark.util.SchemaConvertors

import scala.collection.JavaConverters.asScalaBufferConverter

class DorisThriftReader(partition: DorisReaderPartition, config: DorisConfig) extends AbstractThriftReader(partition, config) {

  private lazy val tableSchema: DorisSchema = {
    val tableIdentifier = config.getValue(DorisOptions.DORIS_TABLE_IDENTIFIER)
    val arr = tableIdentifier.split("\\.")
    frontend.getTableSchema(arr(0), arr(1))
  }

  private lazy val readFields: Array[String] = {
    if (config.contains(DorisOptions.DORIS_READ_FIELDS)) {
      val dorisReadFields = config.getValue(DorisOptions.DORIS_READ_FIELDS)
      if (dorisReadFields == "*") {
        tableSchema.properties.map(_.name.replaceAll("`", "")).toArray
      } else dorisReadFields.split(",").map(_.replaceAll("`", ""))
    } else tableSchema.properties.map(_.name.replaceAll("`", "")).toArray
  }

  private lazy val unsupportedCols = tableSchema.properties
    .filter(field => field.`type`.toUpperCase == "BITMAP" || field.`type`.toUpperCase == "HLL").map(_.name)
    .toArray

  override protected val dorisSchema: DorisSchema = {
    SchemaConvertors.convertToSchema(scanOpenResult.getSelectedColumns.asScala, readFields, unsupportedCols)
  }

}

File: spark-doris-connector/spark-doris-connector-base/src/main/java/org/apache/doris/spark/client/read/DorisThriftReader.java
// Translated from Scala to Java for Apache Doris Spark client
package org.apache.doris.spark.client.read;

import org.apache.doris.spark.client.entity.DorisReaderPartition;
import org.apache.doris.spark.rest.models.Schema;
import org.apache.doris.spark.util.SchemaConvertors;
import scala.collection.JavaConverters;

public class DorisThriftReader extends AbstractThriftReader {

    public DorisThriftReader(DorisReaderPartition partition) throws Exception {
        super(partition);
    }

    @Override
    protected Schema getDorisSchema() {
        return SchemaConvertors.convertToSchema(
                JavaConverters.asScalaBufferConverter(scanOpenResult.getSelectedColumns()).asScala().toSeq());
    }

}